caqti-miou/lib/system.ml

```diff
@@ -0,0 +1,237 @@
+  type t = {
+      mutable value: int
+    ; mutable wakeups: int
+    ; mutex: Miou.Mutex.t
+    ; condition: Miou.Condition.t
+  }
+
+  let create () =
+    {
+      value= 1
+    ; wakeups= 0
+    ; mutex= Miou.Mutex.create ()
+    ; condition= Miou.Condition.create ()
+    }
+
+  let acquire t =
+    Miou.Mutex.protect t.mutex @@ fun () ->
+    while t.value <= 0 do
+      Miou.Condition.wait t.condition t.mutex
+    done;
+    t.value <- t.value - 1
+
+  let release t =
+    Miou.Mutex.protect t.mutex @@ fun () ->
+    t.value <- t.value + 1;
+    Miou.Condition.signal t.condition
+end
+
+module Log = struct
+  type 'a log = 'a Logs.log
+
+  let err ?(src = Logging.default_log_src) = Logs.err ~src
+  let warn ?(src = Logging.default_log_src) = Logs.warn ~src
+  let info ?(src = Logging.default_log_src) = Logs.info ~src
+  let debug ?(src = Logging.default_log_src) = Logs.debug ~src
+end
+
+type stdenv = unit
+
+module Sequencer = struct
+  type 'a t = 'a * Miou.Mutex.t
+
+  let create v = (v, Miou.Mutex.create ())
+  let enqueue (v, m) f = Miou.Mutex.protect m (fun () -> f v)
+end
+
+type _ Effect.t += Spawn : (unit -> unit) -> unit Effect.t
+
+module Switch = struct
+  type hook = (unit -> unit) Miou.Sequence.node
+  type t = { orphans: unit Miou.orphans; hooks: (unit -> unit) Miou.Sequence.t }
+
+  let peek : type a. a Miou.Sequence.t -> a Miou.Sequence.node =
+   fun hooks ->
+    let exception Node of a Miou.Sequence.node in
+    try
+      Miou.Sequence.iter_node ~f:(fun node -> raise (Node node)) hooks;
+      invalid_arg "peek"
+    with Node node -> node
+
+  let release hooks =
+    List.iter (fun fn -> fn ()) (Miou.Sequence.to_list hooks);
+    Miou.Sequence.drop hooks
+
+  let rec terminate orphans =
+    match Miou.care orphans with
+    | None -> ()
+    | Some None -> Miou.yield (); terminate orphans
+    | Some (Some prm) -> (
+        match Miou.await prm with
+        | Ok () -> terminate orphans
+        | Error _exn -> terminate orphans)
+
+  let run fn =
+    let orphans = Miou.orphans () in
+    let hooks = Miou.Sequence.create () in
+    let open Effect.Deep in
+    let retc = Fun.id in
+    let exnc = reraise in
+    let effc : type c. c Effect.t -> ((c, 'r) continuation -> 'r) option =
+      function
+      | Spawn fn ->
+          let _ = Miou.call_cc ~orphans fn in
+          Some (fun k -> continue k ())
+      | _ -> None
+    in
+    let handler = { retc; exnc; effc } in
+    match match_with fn { orphans; hooks } handler with
+    | value -> terminate orphans; release hooks; value
+    | exception exn -> terminate orphans; release hooks; raise exn
+
+  let on_release_cancellable { hooks; _ } fn =
+    Miou.Sequence.(add Left) hooks fn
+
+  let remove_hook hook = Miou.Sequence.remove hook
+
+  let check { orphans; _ } =
+    match Miou.care orphans with
+    | None -> ()
+    | Some None -> ()
+    | Some (Some prm) -> (
+        match Miou.await prm with Ok () -> () | Error _exn -> ())
+end
```

Comment:
Now, I can see you took the effort of implementing a `Switch` specialized for Miou.  Switches are optional, so if they are not part of the suggested coding style of Miou users, then this is not needed.  In particular, the drivers and other Caqti code does not rely on switches for releasing resources it allocates; if it does, it's a bug.

Reply:
Unfortunately, the switch is needed. Miou has some contraints such as only a parent can wait its direct children - more concretely, something like `Lwt.async` is strictly forbidden. The idea of the switch is to aggregate all spawned tasks and be sure that they are _awaited_ by the right parent - otherwise, Miou raises a `Miou.Not_a_child` exception.

This switch implementation is here to ensure that the caqti follows these rules in the way it uses Miou.

Reply:
Aha, I noticed that in the Miou docs.  I'll have a second read with that in mind.  Let me know if you have questions or thoughts on the internal API.  We can make changes there without breaking the public API.